### ISSUE TITLE
Inconsistent use of ServiceProvider.Dispose() in DI specification tests

### ISSUE THREAD

**Issue by lord-executor** [dotnet/runtime#56181]
### Description
[DependencyInjectionSpecificationTests](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/DependencyInjectionSpecificationTests.cs).DisposesInReverseOrderOfCreation is assuming that the `IServiceProvider` implementation also implements `IDisposable` while all other usages indicate that implementing `IDisposable` is optional.

```C#
[Fact]
public void DisposesInReverseOrderOfCreation()
{
    // Arrange
    var serviceCollection = new TestServiceCollection();
    // ...
    var serviceProvider = CreateServiceProvider(serviceCollection);

    // ...

    // Act
    ((IDisposable)serviceProvider).Dispose();

    // Assert
    // ...
}
```

In an ideal world, `IServiceProvider` would probably implement `IDisposable`, but as discussed in #56150 that would constitute a breaking change.

The root of the problem is that `ServiceProvider` is implemented in such a way that it also acts as its own "root scope" without implementing `IServiceScope` but by disposing the scope when the provider is disposed.

This now can mean one of two things:

1. Implementing `IDisposable` on the service provider is really optional.
   In this case, the `DisposesInReverseOrderOfCreation` should be rewritten to use an explicit scope since it tests the order of disposal of instantiated services.
2. Implementing `IDisposable` is not actually optional but instead is needed in order for ASP.NET core to work properly.
   I have not checked that yet. The main purpose of disposing the service provider is (presumably) that instantiated services can be released early, but assuming that it does not directly hold any unmanaged resources the only consequence of not implementing `IDisposable` should be "slightly inefficient use of resources", so I don't think that implementing `IDisposable` is actually required.

I will do some more research and create a PR for variant "1".

### Configuration
* The problem is independent of any configuration

### Other information
If the service provider implementation does not also implement `IDisposable`, then I get
```
Ninject.Web.AspNetCore.ComplianceTests.DependencyInjectionComplianceTests.DisposesInReverseOrderOfCreation
  No source available
   Duration: 97 ms

  Message: 
    System.InvalidCastException : Unable to cast object of type 'Ninject.Web.AspNetCore.NinjectServiceProvider' to type 'System.IDisposable'.
  Stack Trace: 
    DependencyInjectionSpecificationTests.DisposesInReverseOrderOfCreation()
```

**Comment by lord-executor** [comment]
@eerhardt - as discussed in #56150, I created a new bug for this.

**Comment by eerhardt** [comment]
Thanks @lord-executor!

**Comment by davidfowl** [comment]
Implementing IDisposable is pretty much required and it's more than just inefficient use of resources. It's possible to create multiple service providers and leak singleton resources without dispose working. This isn't the common case but libraries pretty much assume as this point that they can dispose the root provider.

> Implementing IDisposable on the service provider is really optional.
In this case, the DisposesInReverseOrderOfCreation should be rewritten to use an explicit scope since it tests the order of disposal of instantiated services.

I'd rather add another test to make sure the root provider implements IDisposable so that the requirement is explicit rather than implicit via this specific disposal test that really tests something else.

**Comment by lord-executor** [comment]
Ah, I see. That would of course also work and make "my type of problem" much easier to debug because there is a clear test that _specifies_ the assumptions behind the service provider.

Would it make sense then to remove the conditional disposal from the other test case(s) or would it be better to still keep them conditional outside of that one test that specifically tests for `IDisposable`?

**Comment by davidfowl** [comment]
> Would it make sense then to remove the conditional disposal from the other test case(s) or would it be better to still keep them conditional outside of that one test that specifically tests for IDisposable?

I'd prefer this.

**Comment by lord-executor** [comment]
StructureMap, or more specifically [StructureMap.Microsoft.DependencyInjection](https://github.com/structuremap/StructureMap.Microsoft.DependencyInjection) does not implement `IDisposable` on its service provider which is why it is skipping the one test that did unconditional disposal (see `StructureMapDependencyInjectionSpecificationTests`).

Since that library is no longer being maintained and has been superseedes by Lamar, I think I will just extend the list of skipped tests to include all other tests that involve disposing the service provider so that we can always call dispose and not have "partially successful" tests which is what is happening now.

**Comment by lord-executor** [comment]
Just realized... that was a bit unclear from me and/or I may have misunderstood the response. @davidfowl _which_ option do you prefer?
1. Remove the conditional disposal from the other test case(s) - making them _unconditional_
2. Keep them conditional outside of that one test that specifically tests for IDisposable

**Comment by davidfowl** [comment]
What you did was good 👍🏾

**Comment by lord-executor** [comment]
OK, weird question @davidfowl : Is it actually functionally relevant that the instantiated services are disposed in the exact reverse order in which they were created? In Ninject, instances are disposed once their associated scope is disposed, so in the multi-inject scenario of `DisposesInReverseOrderOfCreation`, guaranteeing that is going to be a bit difficult.

Given that singletons usually live much longer than anything else, implementing a service that relies on all of its dependencies being disposed before the service itself is disposed seems... dangerous. And if I _did_ need to rely on that then I would probably dispose my dependencies manually in the dispose method of that service. As long as all the services that should be disposed are actually disposed, the order shouldn't matter and if it does, then relying on the DI container to do so seems weird to me.

Superficially, that test just looks like it test a particular implementation detail of the `ServiceProvider`, but that might very well be because I just can't think of a use case where that feature would be relevant. Any hints would be greatly appreciated.

**Comment by davidfowl** [comment]
Each of these tests are there for a reason. You can skip the test for your provider if it can't pass and its likely something will be broken in some subtle way. It's hard to say what though (I can't remember which component rely on that off hand).

**Comment by lord-executor** [comment]
Pity. I was hoping you would know. I like to document this kind of stuff in my code, but I guess I'll just refer to the specification tests there. I managed to find a somewhat reasonable implementation to do disposal in reverse order yesterday, so I hope that I'll never have to find those cases _the hard way_ ;)

**Comment by eerhardt** [comment]
Resolved by #56227.
